File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/CommandRouter.java
package ro.razvan.java.dagger.atm;

import ro.razvan.java.dagger.atm.Command.Result;

import javax.inject.Inject;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public final class CommandRouter {

    private final Map<String, Command> commands;
    private final Outputter outputter;

    @Inject
    public CommandRouter(Map<String, Command> commands, Outputter outputter) {
        this.commands = commands;
        this.outputter = outputter;
    }

    Result route(final String input) {
        final List<String> splitInput = split(input);
        if (splitInput.isEmpty()) {
            return invalidCommand(input);
        }

        final var commandKey = splitInput.get(0);
        final var command = commands.get(commandKey);

        if (command == null) {
            return invalidCommand(input);
        }

        final var result = command.handleInput(splitInput.subList(1, splitInput.size()));
        if (result.isInvalid()) {
            outputter.output(commandKey + ": invalid arguments");
        }

        return result;
    }

    private Result invalidCommand(final String input) {
        outputter.output(String.format("Could't understand \"%s\". Please try again.", input));
        return Result.invalid();
    }

    private static List<String> split(final String input) {
        return Arrays.asList(input.split("\\s+"));
    }
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/CommandRouter.kt
@file:Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")

package ro.razvan.kotlin.dagger.atm

import ro.razvan.kotlin.dagger.atm.Command.Status
import java.util.Map
import javax.inject.Inject

class CommandRouter @Inject constructor(
    private val commands: Map<String, Command>,
    private val outputter: Outputter
) {

    fun route(input: String): Status {
        val splitInput = input.split()
        if (splitInput.isEmpty()) {
            return invalidCommand(input)
        }

        val commandKey = splitInput[0]
        val command = commands[commandKey] ?: return invalidCommand(input)

        val status = command.handleInput(splitInput.subList(1, splitInput.size))

        if (status == Status.INVALID) {
            outputter("$commandKey: invalid arguments")
        }

        return status
    }

    private fun invalidCommand(input: String): Status {
        outputter("Couldn't understand \"$input\". Please try again.")
        return Status.INVALID
    }
}

private fun String.split() = split(regex = Regex.fromLiteral(" "))